lib/Conversion/ImportVerilog/Structure.cpp

```diff
@@ -688,6 +722,31 @@ Context::convertModuleHeader(const slang::ast::InstanceBodySymbol *module) {
       return {};
     }
   }
+
+  // Mapping the hierarchical names into the module's ports.
+  for (auto &hierPath : hierPaths[module]) {
+    auto hierLoc = convertLocation(hierPath.first->location);
+    auto hierType = convertType(hierPath.first->getType());
+
+    // TODO: Skip the hierarchical name regarded as the inout port. Such as
+    // `Top.sub.var = x1` and `x2 = Top.sub.var` both exist in the same region.
+    if (hierPath.second.importPath && hierPath.second.exportPath) {
+      mlir::emitError(hierLoc,
+                      "unsupport tagging a hierarchical name as an inout port");
+      return {};
+    }
+
+    if (auto hierName = hierPath.second.exportPath) {
+      hierPath.second.idx = outputIdx++;
+      modulePorts.push_back({*hierName, hierType, hw::ModulePort::Output});
+    }
+
+    if (auto hierName = hierPath.second.importPath) {
+      hierPath.second.idx = inputIdx++;
+      modulePorts.push_back({*hierName, hierType, hw::ModulePort::Input});
+      block->addArgument(hierType, hierLoc);
+    }
```

Comment:
If I understand correctly, a hiername is marked as `exportPath` or `importPath` depending on whether it is used as an lvalue or rvalue (or both). This means you can get a hiername that is both imported and exported if some module both reads _and_ writes it.

I think you can get rid of that bidirectionality by _always_ passing variables targeted by hiernames around as ref types. So when a module reads and/or writes a variable, you could make the `ref<T>` of that variable available in the module (either as an input port of `ref<T>` type, or as an output port of a submodule of `ref<T>` type). Then you can both read and write from that ref.

A side-effect of this would be that a module never both imports _and_ outputs a variable at the same time. Instead, a module imports a variable ref if it is needed for a hiername locally or in a submodule, and it exports a variable ref if it is used in a hiername in one of the parent modules. But never both. So instead of having separate `exportPath` and `importPath` fields, your hierinfo could just contain a `direction` field that indicates whether it is being imported or exported by a module 😃. That would allow you to support all the weird use cases of hiernames! And we can add an optimization pass later on that converts `ref<T>` ports to `T` ports where possible.

Reply:
Yeah, you're right! The `importPath` corresponds to an rvalue and `exportPath` corresponds to an lvalue, if these two exist, it will be marked as the ref port.

Great idea! Mark all hierarchical names as `ref` ports, and introduce a `direction` to indicate whether it's being imported or exported by a module. I'll apply this thought 🚀 !

Reply:
I don't add a `direction` field to indicate the direction of a hierarchical name. We don't have to use it when we create a module or an instance. WDYT @fabianschuiki?

Reply:
In order to collect all hierarchical names, we traverse the whole AST detailedly. The same traversing exists when creating Moore ops. Hierarchical names are involved everywhere, such as `foo.a = foo.arr[foo.b]`(maybe this case will be nested in `for{ if{always...}}`). So we must add different `visit(expr/stmt/node)`s for different statements, expressions, etc, which is similar to creating Moore ops.